### DISCUSSION TITLE
Unify pg, sqlite, mysql and mssql connection

### DISCUSSION THREAD

**Discussion by khalyomede** [vlang/v#11876] [Ideas]
## Description

Currently, this is the return type of the `connect()` function for these 4 drivers:

```v
// mssql
fn (mut conn Connection) connect(conn_str string) ?bool

// mysql
fn (mut conn Connection) connect() ?bool

// pg
fn connect(config Config) ?DB

// sqlite
fn connect(path string) ?DB
```

For package makers, this can get tought to support different ways to connect to these DB. One that would want to take in parameter a DB connection should hand 4 cases.

## Proposals

A unification of the return type of the `connect()` method for both of the SQL drivers

```v
// mssql
fn connect (config MssqlConfig) ?DB

// mysql
fn connect (config MysqlConfig) ?DB

// pg
fn connect (config PgConfig) ?DB

// sqlite
fn connect (config SqliteConfig) ?DB
```

Each `MysqlConfig`, `PgConfig`, ... structs hold some driver specific parameters.

The `DB` structs becomes a common struct, that hold the same functions (`DB.exec()`, `DB.exec_one()`, ...) for all the drivers.

Here is what it might look like in a "real" example:

```v
import mysql
import sqlite
import mssql
import pg

fn main() {
  mut db := mysql.connect({
    host: "localhost",
    port: 3306,
    user: "root",
    password: "root",
    database: "foo"
  }) or {
    panic(err)
  }

  db = mssql.connect({
    host: "localhost",
    port: 1433,
    user: "root",
    password: "root",
    database: "foo"
  }) or {
    panic(err)
  }

  db = pg.connect({
    host: "localhost",
    port: 5432,
    user: "root",
    password: "root",
    database: "foo"
  }) or {
    panic(err)
  }

  db := sqlite.connect({
    path: "database.sqlite3",
    memory: false // if true, create an in-memory database, could be a nice to have in the future
  }) or {
    panic(err)
  }
}
```

@rcorre correctly suggested we also should be able to connect through an URI. I propose to use a separated function that will parse an URI, extracting the connections parameters, and pass this connection string directly to the driver:

```v
import pg

fn main() {
  db := pg.connect_uri("postgres:host=localhost;port=5432;dbname=foo;user=root") or {
    panic(err) 
  } 
} 
```

The DB struct become a centralized struct (maybe put it in `sql` lib?). It should hold a "driver" property, for the struct to know with which program it should execute (`db.exec()`, `db.exec_one()`) the query. This will help package developers to take into parameter a single type of db instead of dealing with 4:

```v
import sql { DB, Row }

struct Query {
  db DB [required]
}

fn (query Query) all(table string) ?[]Row {
  // Here, handle driver speicifc differences when making the query, like quotes etc...
  match db.driver {
    .mysql {}
    .sqlite {}
    .pg {}
    .mssql {}
  }

  return db.exec("SELECT * FROM $table")
}

fn main() {
  db := sqlite.connect({ path: "database.sqlite3" }) or { panic(err) }

  query := Query{
    db: db
  }

  products := query.all("products") or { panic(err) }
}
```

**Comment by shadowninja55** [comment]
good suggestion, i'll start work on this

**Comment by shadowninja55** [comment]
sorry, i don't think i know the orm code well enough to do this one. i'll leave it up for someone else

**Comment by JalonSolov** [comment]
@LouisSchmieder has been working on a redesign of ORM, with the goal of making a common interface.

**Comment by rcorre** [comment]
It would be really nice if you could also connect to a database via a connection URI instead of a set of parameters. If you're running in e.g. heroku, you get a `$DATABASE_URL` environment variable. At the moment, you have to parse that into its components and build out a struct, which `connect` then assembles _back into the string you started with_.

Anyways, @LouisSchmieder hope you can work that into your redesign somehow :)

Or if the redesign is further out, would folks object to allowing uri-based-connections somehow in the current interface?

Here's a workaround for now if anyone else needs this:

```v
	uri := urllib.parse(connstr) ?
	if uri.scheme != 'postgresql' {
		return error('invalid scheme $uri.scheme in $uri')
	}
	host_and_port := uri.host.split(':')
	return pg.connect(
		host: host_and_port[0]
		port: if host_and_port.len > 1 { host_and_port[1].int() } else { 5432 }
		user: uri.user.username
		password: uri.user.password
		dbname: uri.path.trim('/')
	)
```

**Comment by rcorre** [comment]
> @rcorre correctly suggested we also should be able to connect through an URI. I propose to use a separated function that will parse an URI, extracting the connections parameters, and pass it to the connect function of the given driver:

Sounds good! Only thing I'll add is that, at least in the case of postgres, you shouldn't have to parse anything. You can hand the string directly to `PQconnectdb`.

**Comment by dumblob** [comment]
> Sounds good! Only thing I'll add is that, at least in the case of postgres, you shouldn't have to parse anything. You can hand the string directly to `PQconnectdb`.

I'd actually oppose this as it'd make debugging etc. a nightmare. If any unification (which this discussion is about AFAIK), then the behavior (incl. parsing if any) has to be **identical for all backends** (because guaranteed *V* will parse the string differently than PQconnectdb).

**Comment by rcorre** [comment]
> because guaranteed V will parse the string differently than PQconnectdb

V won't parse strings. The end-user calls `[pg,sqlite,mysql,mssql]_connect_uri(uri)`, and the string `uri` is passed unchanged to the native `connect` call, similar to how [`db.Open`](https://golang.org/pkg/database/sql/#Open) works in Go.

**Comment by dumblob** [comment]
Ah, ok. Then it's fine.

Btw. why did you then emphasize `PQconnectdb` and not others? That's why I initially understood that for others *V* shall parse it :wink:.

**Comment by rcorre** [comment]
> why did you then emphasize PQconnectdb and not others?

Sorry, that was just an example because I was working with postgres at the time :)

Though at a quick glance, maybe the mysql API doesn't support connection strings directly? I'm not very familiar with mysql/mssql.

**Comment by khalyomede** [comment]
FYI, MySQL and MSSQL both support connecting using an URI:
- https://dev.mysql.com/doc/refman/8.0/en/connecting-using-uri-or-key-value-pairs.html
- https://docs.microsoft.com/en-US/biztalk/adapters-and-accelerators/adapter-sql/create-the-sql-server-connection-uri

Speaking of tech design, I think what should all `connect()` method do is simply generating the correct connection string to pass to the database server down the line. This way these methods will call `connect_uri()` behind the scene, and they act as a nicer API to perform the connection.

Looks like using a connection string is quite common for database engines (at least for the 4 mentioned in this issue). We can see other "less popular" (all proportion saved) database engines use the same mecanism, enough to think this pattern is safe to follow:

- https://dbschema.com/documentation/Cassandra/

**Comment by JalonSolov** [comment]
URI is the way Java has been doing this for years, in their database layer.  The Java code handles whether to just pass the URI along, or break it up as needed.

**Comment by rcorre** [comment]
> Speaking of tech design, I think what should all connect() method do is simply generating the correct connection string to pass to the database server down the line

FWIW, that's already how postgres is written: 
https://github.com/vlang/v/blob/bafea57622ae3ffba9316a990876a4c3cec9c143/vlib/pg/pg.v#L78-L79
